pkg/controllers/management/etcdbackup/etcdbackup.go

```diff
@@ -0,0 +1,298 @@
+	cluster, err := c.clusterClient.Get(b.ClusterID, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig == nil {
+		return nil, fmt.Errorf("[etcd-backup] cluster doesn't have a backup config")
+	}
+
+	if !v3.BackupConditionCreated.IsTrue(b) {
+		b.Filename = getBackupFilename(b.Name, cluster)
+		b.BackupConfig = *cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig
+		v3.BackupConditionCreated.True(b)
+		b, err = c.backupClient.Update(b)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	kontainerDriver, err := c.KontainerDriverLister.Get("", service.RancherKubernetesEngineDriverName)
+	if err != nil {
+		return nil, err
+	}
+	bObj, saveErr := v3.BackupConditionCompleted.Do(b, func() (runtime.Object, error) {
+		err = c.backupDriver.ETCDSave(context.Background(), cluster.Name, kontainerDriver, cluster.Spec, b.Name)
+		return nil, err
+	})
+	b, err = c.backupClient.Update(bObj.(*v3.EtcdBackup))
+	if err != nil {
+		return nil, err
+	}
+	if saveErr != nil {
+		return nil, fmt.Errorf("[etcd-backup] failed to perform etcd backup: %v", saveErr)
+	}
+	return b, nil
+}
+
+func (c *Controller) Remove(b *v3.EtcdBackup) (runtime.Object, error) {
+	logrus.Infof("[etcd-backup] Deleteing backup %s ", b.Name)
+	return b, c.deleteS3Snapshot(b)
+}
+
+func (c *Controller) Updated(b *v3.EtcdBackup) (runtime.Object, error) {
+	return b, nil
+}
+
+func (c *Controller) clusterBackupSync(ctx context.Context, interval time.Duration) error {
+	for range ticker.Context(ctx, interval) {
+		clusters, err := c.clusterLister.List("", labels.NewSelector())
+		if err != nil {
+			logrus.Error(fmt.Errorf("[etcd-backup] clusterBackupSync faild: %v", err))
+			return err
+		}
+		for _, cluster := range clusters {
+			logrus.Infof("[etcd-backup] Checking backups for cluster: %s", cluster.Name)
+			if err := c.doClusterBackupSync(cluster); err != nil && !apierrors.IsConflict(err) {
+				logrus.Error(fmt.Errorf("[etcd-backup] clusterBackupSync faild: %v", err))
+			}
+		}
+	}
+	return nil
+}
+
+func (c *Controller) doClusterBackupSync(cluster *v3.Cluster) error {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil
+	}
+	// check if the cluster is eligable for backup.
+	if !shouldBackup(cluster) {
+		return nil
+	}
+
+	clusterBackups, err := c.backupLister.List(cluster.Name, labels.NewSelector())
+	if err != nil {
+		return err
+	}
+
+	// cluster has no backups, we need to kick a new one.
+	if len(clusterBackups) == 0 {
+		logrus.Infof("[etcd-backup] Cluster [%s] has no backups, creating first backup", cluster.Name)
+		newBackup, err := c.createNewBackup(cluster)
+		if err != nil {
+			return fmt.Errorf("[etcd-backup] Backup create failed:%v", err)
+		}
+		logrus.Infof("[etcd-backup] Cluster [%s] new backup is created: %s", cluster.Name, newBackup.Name)
+		return nil
+	}
+
+	newestBackup := clusterBackups[0]
```

Comment:
would this break if `clusterBackups` is empty? 
in that case, let's assign newestBackup in the logic where `len(clusterBackups) == 0`